src/api/homa.ts

```diff
@@ -0,0 +1,135 @@
+import { ContractTransaction } from 'ethers';
+import { DOT } from '@acala-network/contracts/utils/AcalaTokens';
+import { ERC20__factory, HomaFactory__factory } from '@acala-network/asset-router/dist/typechain-types';
+import { KSM } from '@acala-network/contracts/utils/KaruraTokens';
+
+import {
+  Mainnet,
+  RouteError,
+  RouteParamsHoma,
+  _populateRelayTx,
+  _populateRouteTx,
+  getChainConfig,
+  getMainnetChainId,
+  routeStatusParams,
+  toAddr32,
+} from '../utils';
+
+const prepareRouteHoma = async (chain: Mainnet) => {
+  const chainId = getMainnetChainId(chain);
+  const chainConfig = await getChainConfig(chainId);
+  const { feeAddr, homaFactoryAddr, wallet } = chainConfig;
+
+  const homaFactory = HomaFactory__factory.connect(homaFactoryAddr!, wallet);
+  const routeToken = chain === Mainnet.Acala ? DOT : KSM;
+
+  return { homaFactory, feeAddr, routeToken, wallet };
+};
+
+export const shouldRouteHoma = async ({ chain, destAddr }: RouteParamsHoma) =>  {
+  try {
+    const { homaFactory, feeAddr } = await prepareRouteHoma(chain);
+    const routerAddr = await homaFactory.callStatic.deployHomaRouter(
+      feeAddr,
+      toAddr32(destAddr),
+    );
+
+    return {
+      shouldRoute: true,
+      routerAddr,
+    };
+  } catch (err) {
+    return {
+      shouldRoute: false,
+      msg: err.message,
+    };
+  }
+};
+
+export const routeHoma = async ({ chain, destAddr }: RouteParamsHoma) =>  {
+  const { homaFactory, feeAddr, routeToken } = await prepareRouteHoma(chain);
+  const tx = await homaFactory.deployHomaRouterAndRoute(feeAddr, toAddr32(destAddr), routeToken);
+  const receipt = await tx.wait();
+
+  return receipt.transactionHash;
+};
+
+/* --------------------------------- auto route homa --------------------------------- */
+export enum RouteStatus {
+  Waiting = 0,
+  Routing = 1,
+  Confirming = 2,
+  Complete = 3,
+  Timeout = -1,
+  Failed = -2,
+};
+
+interface RouteInfo {
+  status: RouteStatus;
+  [key: string]: any;
+}
+
+let routeReqId = 0;
```

Comment:
this means id will reset after server restart.
how about just make it start from `Date.now()`

Reply:
yeah that's a good idea, I will make `const reqId = `homa-${Date.now()}`;` so each id is unique and contains info about when it is received

Reply:
currently doing a couple small polishes, will commit the changes together

Reply:
only using timestamp is not good because it could have collections. so either add another sequential number or a random string

Reply:
ok let's make seqId starts from Date.now()